### PR TITLE
feat(set_admin): new groups, chasepull and aicontrol

### PR DESCRIPTION
**Pull request by dd84ai**
feat(set_admin): adds four new subgroups to set admin command

setadmin PlayerName chasepull, which allows
```
chase
pull
```
setadmin PlayerName aicontrol, which allows
```
aidestroy
aicancel
aifollow
aicome
aifleet
fleetlist
```
setadmin PlayerName cloak, which allows
```
cloak
```
setadmin PlayerName bases, which allows
```
testrecipe
testdeploy
basedestroy
basetogglegod
testbase
jumpcreate
basecreate
basedebugon
basedebugoff
```
Additionally:
locked commands in pvecontroller to be under
setadmin PlayerName plugins
```
pvecontroller <command>
```
and some weird commands from player_restrictions
made sure to hide(or rehide) under superadmin rights
```
showrestrictions
generateid
shiptest
healthtest1
healthtest2
testfuseobj
testunfuseobj
testselffuseobj
testselfunfuseobj
```

### CODE REVIEW

**Comment by dd84ai** [comment]
btw, tested in game that everything works fine:
chasepull admin is not able to do aicreate
and aiecreate admin is not able to chasepull
and both can't beam

**Comment by dd84ai** [comment]
Do not merge, I have a problem

**Comment by dd84ai** [comment]
Solved everything

**Comment by dd84ai** [comment]
Performed tests:
]
1) setadmin Player cloak,chasepull, 
it can cloak, it can chase, but can't setcash and can't beam Erie and can't basecreate
2) setadmin Player aicontrol,beamkill
it can't cloak, it can aicreate, it can't setcash, it can beam
3) setadmin Player bases,beamkill
it can basecreate, it can beamkill, it can't cloak, it can't setcash, it can't use .generateid
4) setadmin Player superadmin
it can use generated, it can cloak, it can chasepull, it can basecreate, it can pvecontroller status
5) setadmin Player plugins,beamkill
it can pvecontroller status, it can't cloak, it can beam

All tests have been passed

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 1]
```diff
@@ -768,31 +768,34 @@ bool UserCmd_Process(uint iClientID, const wstring &wscCmd)
...
+
+		RIGHT_CHECK(RIGHT_SUPERADMIN)
+
 		return true;
 	}
 	else if (IS_CMD("generateid"))
 	{
 		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+
+		RIGHT_CHECK(RIGHT_SUPERADMIN)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:788]: this is actually harmless and should probably not even be an admincmd - actually I would question why it's an FLHook command at all, and not some out-of-game utility. But this is the existing check so okay 👍

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 2]
```diff
@@ -768,31 +768,34 @@ bool UserCmd_Process(uint iClientID, const wstring &wscCmd)
...
+#define RIGHT_CHECK(a) if(!(cmds->rights & a)) { cmds->Print(L"ERR No permission\n"); return true; }
 bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
 {
 	returncode = DEFAULT_RETURNCODE;
 
 	if (IS_CMD("showrestrictions"))
 	{
 		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+
+		RIGHT_CHECK(RIGHT_SUPERADMIN)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:780]: this command... does nothing???

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 3]
```diff
@@ -768,31 +768,34 @@ bool UserCmd_Process(uint iClientID, const wstring &wscCmd)
...
 	{
 		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
 		HKPLAYERINFO adminPlyr;
-		if (HkGetPlayerInfo(cmds->GetAdminName(), adminPlyr, false) != HKE_OK)
-		{
-			cmds->Print(L"ERR\n");
-			return true;
-		}
+		
+		RIGHT_CHECK(RIGHT_SUPERADMIN)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:798]: this is actually harmless and I question whether it should be admin-restricted, though I also wonder whether it should really be an FLHook command at all.

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 4]
```diff
@@ -923,11 +912,7 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
 	}
 	else if (IS_CMD("testselffuseobj"))
 	{
-		if (cmds->rights != RIGHT_SUPERADMIN)
-		{
-			cmds->Print(L"ERR No permission\n");
-			return true;
-		}
+	RIGHT_CHECK(RIGHT_SUPERADMIN)
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:915]: missing indentation

File: Plugins/Public/npc_control/Main.cpp [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #include <sstream>
 #include <iostream>
 
+#define RIGHT_CHECK(a) if(!(cmds->rights & a)) { cmds->Print(L"ERR No permission\n"); return; }
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp:29]: should we just define this thing straight into FLHook core itself?

**Review by dsyalex — CHANGES_REQUESTED** [review]
reee indentation and code duplication. otherwise looks fine :)

**Review by dsyalex — CHANGES_REQUESTED** [review]
I have a feeling those bitwise-style checks in RIGHT_CHECK will fail against RIGHT_SUPERADMIN

If you consider it as 8 bits for this example:
Let's say you have rights 00010000
RIGHT_SUPERADMIN in our example is 11111111
check is `!(cmds->rights & RIGHT_SUPERADMIN)` to find unauthorised access
00010000 & 11111111
result: 00010000
!00010000
result: false
I think it breaks

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 6]
```diff
@@ -768,29 +762,31 @@ bool UserCmd_Process(uint iClientID, const wstring &wscCmd)
...
+			return true;
+		}
+
 		AdminCmd_GenerateID(cmds, cmds->ArgStrToEnd(1));
 		return true;
 	}
 	else if (IS_CMD("shiptest"))
 	{
 		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
 		HKPLAYERINFO adminPlyr;
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:785]: can be removed

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 7]
```diff
@@ -803,9 +799,10 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:801]: same here

File: Plugins/Public/alley/PlayerRestrictions.cpp [hunk 8]
```diff
@@ -823,9 +820,10 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
```
Comment (dsyalex) [inline Plugins/Public/alley/PlayerRestrictions.cpp:822]: and here

File: Plugins/Public/npc_control/Main.cpp (hunk 5 above)
Comment (dsyalex) [reply Plugins/Public/npc_control/Main.cpp:29]: A: no, they can be a bit different

**Review by dsyalex — CHANGES_REQUESTED** [review]

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 6 above)
Comment (dsyalex) [reply Plugins/Public/alley/PlayerRestrictions.cpp:785]: nope

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 7 above)
Comment (dsyalex) [reply Plugins/Public/alley/PlayerRestrictions.cpp:802]: nope

File: Plugins/Public/alley/PlayerRestrictions.cpp (hunk 8 above)
Comment (dsyalex) [reply Plugins/Public/alley/PlayerRestrictions.cpp:824]: nope

**Review by dsyalex — APPROVED** [review]

**Pull request merged**